File: packages/expo-bip39/android/src/main/java/expo/modules/bip39/ExpoBip39Module.kt
package expo.modules.bip39

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.CodedException
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.PBEKeySpec
import java.security.NoSuchAlgorithmException
import java.security.spec.InvalidKeySpecException

class ExpoBip39Module : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoBip39')` in JavaScript.
    Name("ExpoBip39")

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("mnemonicToSeed") { mnemonic: String, password: String ->
      deriveSeed(mnemonic, password)
    }
  }

  private fun deriveSeed(mnemonic: String, password: String): List<String> {
    val salt = "mnemonic$password"
    
    try {
      val spec = PBEKeySpec(
        mnemonic.toCharArray(),
        salt.toByteArray(Charsets.UTF_8),
        2048, // 迭代次数
        512   // 输出长度 (bits) - 64 bytes
      )
      
      val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA512")
      val derivedKey = factory.generateSecret(spec).encoded
      
      // 转换为字符串列表
      return derivedKey.map { it.toUByte().toString() }
      
    } catch (e: NoSuchAlgorithmException) {
      throw CodedException("PBKDF2_ERROR", "PBKDF2WithHmacSHA512 algorithm not available", e)
    } catch (e: InvalidKeySpecException) {
      throw CodedException("PBKDF2_ERROR", "Key derivation failed", e)
    } catch (e: Exception) {
      throw CodedException("ENCODING_ERROR", "Invalid input encoding", e)
    }
  }
}

File: packages/expo-bip39/ios/ExpoBip39Module.swift
import ExpoModulesCore
import CommonCrypto

public class ExpoBip39Module: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoBip39')` in JavaScript.
    Name("ExpoBip39")

    AsyncFunction("mnemonicToSeed") { (mnemonic: String, password: String) in
      return try self.deriveSeed(mnemonic: mnemonic, password: password)
    }
  }

  private func deriveSeed(mnemonic: String, password: String) throws -> [String] {
    let salt = "mnemonic" + password
    guard let passData = mnemonic.data(using: .utf8),
          let saltData = salt.data(using: .utf8) else {
      throw Exception(name: "ENCODING_ERROR", description: "Invalid input encoding")
    }

    // 准备输出缓冲区 111
    var derivedKey = [UInt8](repeating: 0, count: 64)

    // 调用 CommonCrypto PBKDF2-HMAC-SHA512
    let status = passData.withUnsafeBytes { passBytes in
      saltData.withUnsafeBytes { saltBytes in
        CCKeyDerivationPBKDF(
          CCPBKDFAlgorithm(kCCPBKDF2),         // 算法
          passBytes.bindMemory(to: Int8.self).baseAddress, passData.count,  // 密码及长度
          saltBytes.bindMemory(to: UInt8.self).baseAddress, saltData.count, // salt 及长度
          CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA512),
          2048,                                 // 迭代次数
          &derivedKey, derivedKey.count        // 输出缓冲区
        )
      }
    }

    guard status == kCCSuccess else {
      throw Exception(name: "PBKDF2_ERROR", description: "Key derivation failed")
    }

    // 转换为字符串数组
    return derivedKey.map { String($0) }
  }
}
